### PR TITLE
182867738 enrollment mutation

### PR DESCRIPTION
**Pull request by ZachCMP**
- Add `CreateEnrollment` mutation
- Add `CreateEnrollmentValues` input type
- Adjust `Enrollment` type to add new fields and fix dates
- Add `ErrorGroup` object to handle passing groups of related errors
- Update `ValidationError` type to handle more unpredictable values
- Add WIP models & mechanism for enrollments
- Add tests for the mutation

Couple questions:

1. Since we're requiring all of the fields in the input type that the enrollment itself requires to be valid, it kind of seems to me that nothing could be created from that input that would be "in progress" due to being structurally invalid, but maybe it would be in progress for a different reason. Is that correct?

2. Depending on the answer to the above, we'll probably need to add more test cases to test the WIP workflow

### CODE REVIEW

File: drivers/hmis/app/graphql/schema.graphql
Comment (gigxz) [inline drivers/hmis/app/graphql/schema.graphql:185]: should this be `enrollmentMutationId` or just `mutationId`? "client" doesnt seem right

File: drivers/hmis/app/graphql/mutations/create_enrollment.rb
Comment (gigxz) [inline drivers/hmis/app/graphql/mutations/create_enrollment.rb]: I think we should only save the Enrollment as "in progress" if the client explicitly requests that by passing `in_progress: true`. So if the client tries to submit an Enrollment that is invalid, we should be returning validation errors instead of silently saving as in-progress.

File: drivers/hmis/app/graphql/types/hmis_schema/create_enrollment_values.rb
Comment (gigxz) [inline drivers/hmis/app/graphql/types/hmis_schema/create_enrollment_values.rb]: we should validate that the project exists. we should probably only look for the project within the viewably_by scope for the current user, so we're not giving them access to create enrollments in projects that they shouldn't be able to see.

File: drivers/hmis/app/graphql/mutations/create_enrollment.rb
Comment (gigxz) [inline drivers/hmis/app/graphql/mutations/create_enrollment.rb]: This feels like something that could be written as a validator on `argument :start_date`? It doesn't look like they have any [built-ins](https://graphql-ruby.org/fields/validation.html#built-in-validations) for date, so it would need to be custom. Seems worthwhile; we are going to need to validate date inputs on other mutations too (event, service..).

Comment (eanders) [inline drivers/hmis/app/graphql/mutations/create_enrollment.rb]: This looks like it's only validating one field at a time.  Should it maybe have an `add_error` method and do all the validations before returning?

File: drivers/hmis/app/graphql/mutations/create_enrollment.rb
Comment (gigxz) [inline drivers/hmis/app/graphql/mutations/create_enrollment.rb:67]: I'm not sure how I feel about the shape of `errors` changing. Maybe instead of this, we could attach the `id` to each error inside `enrollment.errors`?  I could be convinced either way , I think.

File: drivers/hmis/app/graphql/schema.graphql
Comment (ZachCMP) [reply drivers/hmis/app/graphql/schema.graphql:185]: `client` here is referring to the browser-side of the application, not a HUD client. Also you'd never really deal with these arguments in practice. We can probably remove them without issue by modifying `base_mutation`

Comment (gigxz) [reply drivers/hmis/app/graphql/schema.graphql:185]: oh okay, I assumed it was about the Client since that's the only mutation we have :) I thought it might be a copy-paste error. sounds good

File: drivers/hmis/app/graphql/mutations/create_enrollment.rb
Comment (ZachCMP) [reply drivers/hmis/app/graphql/mutations/create_enrollment.rb]: @eanders What do you think about that? The above is what we discussed, but I can see arguments either way. Maybe we should all discuss together?

Comment (eanders) [reply drivers/hmis/app/graphql/mutations/create_enrollment.rb]: I could go either way.  This felt safer, the errors are still surfaced, but the data gets saved.

File: drivers/hmis/app/graphql/mutations/create_enrollment.rb
Comment (ZachCMP) [reply drivers/hmis/app/graphql/mutations/create_enrollment.rb:67]: Yeah, in retrospect I think I'd prefer to have an ID in each error. The tricky bit is that the actual objects backing the `ValidationError` type are `ActiveModel::Error` objects, so I can't just append an ID key to them and call it a day. But I'll take another look and see if I can change it.

File: drivers/hmis/app/graphql/mutations/create_enrollment.rb
Comment (ZachCMP) [reply drivers/hmis/app/graphql/mutations/create_enrollment.rb]: Yup, definitely. I'll get this fixed up

File: drivers/hmis/app/graphql/mutations/create_enrollment.rb
Comment (gigxz) [reply drivers/hmis/app/graphql/mutations/create_enrollment.rb]: Ok, as long as the validation errors are surfaced, that makes sense to me. I was thinking that the `save(validate: false)` would make it so that the validation errors don't get returned to the client.

File: drivers/hmis/app/graphql/types/hmis_schema/create_enrollment_values.rb
Comment (eanders) [reply drivers/hmis/app/graphql/types/hmis_schema/create_enrollment_values.rb]: I agree!

File: drivers/hmis/app/graphql/mutations/create_enrollment.rb
Comment (ZachCMP) [reply drivers/hmis/app/graphql/mutations/create_enrollment.rb]: Fixed

File: drivers/hmis/app/graphql/types/hmis_schema/create_enrollment_values.rb
Comment (ZachCMP) [reply drivers/hmis/app/graphql/types/hmis_schema/create_enrollment_values.rb]: Fixed

File: drivers/hmis/app/graphql/mutations/create_enrollment.rb
Comment (ZachCMP) [reply drivers/hmis/app/graphql/mutations/create_enrollment.rb:67]: I decided to flatten this and add an ID field to ValidationError, but I still need to get the actual enrollment IDs in there. We might need to think a bit about a pattern for that, since we're passing actual `Error` objects into `errors`

File: drivers/hmis/app/graphql/mutations/create_enrollment.rb
Comment (eanders) [reply drivers/hmis/app/graphql/mutations/create_enrollment.rb]: `validate: false` just means it'll save.  Calling `enrollment.valid?` above will generate the errors.

Comment (gigxz) [reply drivers/hmis/app/graphql/mutations/create_enrollment.rb]: ok! thank you

File: drivers/hmis/spec/requests/hmis/create_enrollment_spec.rb
Comment (gigxz) [inline drivers/hmis/spec/requests/hmis/create_enrollment_spec.rb]: Can you add an entry for validating the `attribute` here? i'm getting `attribute: null` on the entry date validation.

File: drivers/hmis/app/graphql/mutations/create_enrollment.rb
Comment (gigxz) [reply drivers/hmis/app/graphql/mutations/create_enrollment.rb]: Relatedly.. if the enrollment is saved as in progress then the Project should be nil. I'm not seeing that behavior when testing this out with graphiql. I tested by creating an in-progress enrollment and then looking it up in the console: `Hmis::Hud::Enrollment.find(58515).project` should be `nil`.

File: drivers/hmis/spec/requests/hmis/create_enrollment_spec.rb
Comment (gigxz) [inline drivers/hmis/spec/requests/hmis/create_enrollment_spec.rb:99]: Can we add assertions for  `Hmis::Hud::Enrollment.find(id).project`  and/or presence in the `Hmis::Hud::Enrollment.in_progress` scope?

File: drivers/hmis/app/graphql/mutations/create_enrollment.rb
Comment (ZachCMP) [reply drivers/hmis/app/graphql/mutations/create_enrollment.rb]: Should be fixed now

File: drivers/hmis/app/graphql/mutations/create_enrollment.rb
Comment (ZachCMP) [reply drivers/hmis/app/graphql/mutations/create_enrollment.rb]: Added a method to declare date fields (`date_string_argument`) to handle basic format validation. We'll probably want to add more later, but I think this is a good start

File: drivers/hmis/app/graphql/mutations/create_enrollment.rb
Comment (ZachCMP) [reply drivers/hmis/app/graphql/mutations/create_enrollment.rb:67]: Tweaked this a bit to add a custom error type that takes whatever kwargs you give it. Probably not a perfect solution, but it's a start.

File: drivers/hmis/spec/requests/hmis/create_enrollment_spec.rb
Comment (ZachCMP) [reply drivers/hmis/spec/requests/hmis/create_enrollment_spec.rb]: Fixed with the comment above about the custom error type

File: drivers/hmis/spec/requests/hmis/create_enrollment_spec.rb
Comment (ZachCMP) [reply drivers/hmis/spec/requests/hmis/create_enrollment_spec.rb:99]: Yup, done

**Comment by gigxz** [comment]
I'm not seeing the expected behavior regarding WIP records. I would expect the following query to create two WIP Records that have enrollment_id filled in.


```
mutation CreateEnrollment($input: CreateEnrollmentInput!) {
  createEnrollment(input: $input) {
    clientMutationId
    enrollments {
      id
    }
    errors {
      type
      attribute
      message
      fullMessage
    }
  }
}

{
  "input": {
    "projectId": "183",
    "startDate": "2022-09-01",
    "householdMembers": [
      {
        "id": "24440",
        "relationshipToHoH": "SELF_HEAD_OF_HOUSEHOLD_"
      },
      {
        "id": "14172",
        "relationshipToHoH": "OTHER_RELATIVE"
      }
    ],
    "inProgress": true
  }
}
```

Result:
```
{
  "data": {
    "createEnrollment": {
      "clientMutationId": null,
      "enrollments": [
        {
          "id": "58534"
        },
        {
          "id": "58535"
        }
      ],
      "errors": []
    }
  }
}
```

Unexpected result in the rails console below. I would have expected 2 records, both with enrollment_id present. I'm also not sure about why the client_id is 6. I destroyed all the WIP records beforehand, and this is what I'm getting from the above query. Very possible something is wrong with my environemnt... 
```
irb(main):045:0> Hmis::Wip.count
=> 1
irb(main):046:0> Hmis::Wip.last
=> #<Hmis::Wip id: 13, client_id: 6, project_id: 113, enrollment_id: nil, source_type: "Hmis::Hud::Enrollment", source_id: 58535, date: "2022-09-01", data: nil, created_at: "2022-09-06 17:33:26.041728000 -0400", updated_at: "2022-09-06 17:33:26.041728000 -0400", deleted_at: nil>
```

File: drivers/hmis/spec/requests/hmis/create_enrollment_spec.rb
Comment (gigxz) [inline drivers/hmis/spec/requests/hmis/create_enrollment_spec.rb:102]: Can you add assertions for `Wip` records being created with the correct values?

File: drivers/hmis/app/models/hmis/hud/enrollment.rb
Comment (gigxz) [inline drivers/hmis/app/models/hmis/hud/enrollment.rb]: We want the `client.id` here not the personal ID! thanks

**Comment by gigxz** [comment]
And lets remove changes to `db/warehouse_structure.sql` before merging

File: drivers/hmis/spec/requests/hmis/create_enrollment_spec.rb
Comment (ZachCMP) [reply drivers/hmis/spec/requests/hmis/create_enrollment_spec.rb:102]: Done

File: drivers/hmis/app/models/hmis/hud/enrollment.rb
Comment (ZachCMP) [reply drivers/hmis/app/models/hmis/hud/enrollment.rb]: Fixed

**Comment by ZachCMP** [comment]
@eanders I can't seem to remove the `warehouse_structure.sql` changes form this PR. I did `git checkout origin/release-48 db/warehouse_structure.sql` and committed, but it's still showing changes. I might need a hand getting them to go away.

**Comment by gigxz** [comment]
Tested out the WIP record creation, that looks good now 👍 


* I noticed `EnrollmentID` is null on created enrollments (both WIP and non-WIP). Is that intentional? I'm not sure what that field is supposed to represent for hmis-created enrollments. I think typically it's the ID we get from the external HMIS, so maybe we should throw a uuid in there. @eanders would know!

* WIP enrollments aren't appearing in query results yet (`query enrollment(id)` and `query client(id) { enrollments {...} }`). That side of it wasn't explicitly part of this ticket, so I don't think it needs to hold this PR up, we can add a separate ticket for it! Pretty sure that's happening because WIP enrollments get filtered out of `viewable_by` on Enrollment when it joins on project.

edit: created 183226018 to address the last point

**Comment by eanders** [comment]
@ZachCMP , just merged in an updated `release-48` all seems good now.

**Review by eanders — APPROVED** [review]
I defer to @gigxz for final sign-off, but this looks good to me.

File: drivers/hmis/app/graphql/mutations/create_enrollment.rb
Comment (gigxz) [inline drivers/hmis/app/graphql/mutations/create_enrollment.rb:52]: @eanders should we be setting the `EnrollmentID` on this record?

**Review by gigxz — APPROVED** [review]

File: drivers/hmis/app/graphql/mutations/create_enrollment.rb
Comment (eanders) [reply drivers/hmis/app/graphql/mutations/create_enrollment.rb:52]: I believe so.  I can't think of a situation where we wouldn't want that set for Enrollment.

Comment (gigxz) [reply drivers/hmis/app/graphql/mutations/create_enrollment.rb:52]: Ok. Currently it's not being set, so I think that's the last change @ZachCMP. Should we put a uuid in there?

Comment (ZachCMP) [reply drivers/hmis/app/graphql/mutations/create_enrollment.rb:52]: @eanders ^^?

Comment (eanders) [reply drivers/hmis/app/graphql/mutations/create_enrollment.rb:52]: That sounds right to me.  If we don't have an EnrollmentID, we need to set it to a UUID.

Comment (ZachCMP) [reply drivers/hmis/app/graphql/mutations/create_enrollment.rb:52]: This is done now, and is included in the tests

File: drivers/hmis/app/graphql/mutations/create_enrollment.rb
Comment (gigxz) [inline drivers/hmis/app/graphql/mutations/create_enrollment.rb]: The enrollment ID should be unique
```suggestion
          enrollment_id: SecureRandom.uuid.gsub(/-/, ''),
```

Comment (ZachCMP) [reply drivers/hmis/app/graphql/mutations/create_enrollment.rb]: @gigxz Resolved

**Review by gigxz — APPROVED** [review]

**Pull request merged**
